File: happy-ladybugs/happy-ladybugs-2.c
#include <assert.h>
#include <stdint.h>
#include <stdio.h>

int happy_ladybug(const char* b) {
  uint8_t d[26] = {0};
  uint8_t has_empty = 0;
  uint8_t already_happy = 1;
  uint8_t adj_count = 0;
  uint8_t left_c = 0;

  while (*b != 0) {
    uint8_t c = *b;

    if (c == '_') {
      has_empty = 1;
    } else {
      d[c - 'A'] += 1;

      if (c == left_c) {
        adj_count += 1;
      }

      if ((adj_count < 2) && (left_c != 0)) {
        already_happy = 0;
      }

      if (c != left_c) {
        adj_count = 1;
      }

      left_c = c;
    }

    ++b;
  }

  if ((adj_count < 2) && (left_c != 0)) {
    already_happy = 0;
  }

  uint8_t all_at_least_two = 1;

  for (uint8_t i = 0; i < 26; ++i) {
    if (d[i] == 1) {
      all_at_least_two = 0;
    }
  }

  return already_happy || (has_empty && all_at_least_two);
}

int main() {
  assert(happy_ladybug("_") == 1);
  assert(happy_ladybug("AA") == 1);
  assert(happy_ladybug("A_BBA") == 1);
  assert(happy_ladybug("_ABBA") == 1);
  assert(happy_ladybug("AABB") == 1);
  assert(happy_ladybug("__AA") == 1);
  assert(happy_ladybug("BB__AA") == 1);
  assert(happy_ladybug("BB_AA") == 1);
  assert(happy_ladybug("BBAA_") == 1);
  assert(happy_ladybug("A") == 0);
  assert(happy_ladybug("_A") == 0);
  assert(happy_ladybug("AABBCD") == 0);
  assert(happy_ladybug("ABAB") == 0);
  assert(happy_ladybug("X_") == 0);
  assert(happy_ladybug("X_Y_Z_") == 0);
  assert(happy_ladybug("AABBC") == 0);
}

// def main():

// main()

File: happy-ladybugs/happy-ladybugs-2.py
#!/bin/python3

def happy_ladybug(b):
    d = {}
    has_empty = False
    already_happy = True
    adj_count = 0
    left_c = None

    for c in b:
        if c == '_':
            has_empty = True
        else:
            d[c] = d.setdefault(c, 0) + 1

            if c == left_c:
                adj_count += 1

            if (adj_count < 2) and (left_c is not None):
                already_happy = False

            if c != left_c:
                adj_count = 1

            left_c = c

    if (adj_count < 2) and (left_c is not None):
        already_happy = False

    all_at_least_two = all([x >= 2 for x in d.values()])

    return already_happy or (has_empty and all_at_least_two)


def main():
    assert happy_ladybug("_")
    assert happy_ladybug("AA")
    assert happy_ladybug("A_BBA")
    assert happy_ladybug("_ABBA")
    assert happy_ladybug("AABB")
    assert happy_ladybug("__AA")
    assert happy_ladybug("BB__AA")
    assert happy_ladybug("BB_AA")
    assert happy_ladybug("BBAA_")
    assert not happy_ladybug("A")
    assert not happy_ladybug("_A")
    assert not happy_ladybug("AABBCD")
    assert not happy_ladybug("ABAB")
    assert not happy_ladybug("X_")
    assert not happy_ladybug("X_Y_Z_")
    assert not happy_ladybug("AABBC")


main()
